### DISCUSSION TITLE
Is there a possibility for a library user to make size/perf choices?

### DISCUSSION THREAD

**Discussion by VeryOldFrank** [Bodmer/TFT_eSPI#2057] [Q&A - Fonts]
I'm using this library on an Arduino WIO. 
I'm displaying a few (sometimes rapidly) changing values on the WIO screen in size 4 font. 
Whenever a value changes, I overwrite the previous in white color font and the new one in black color font. (This is faster than drawing a white rectangle over the previous field) 
I understand from the .h file, that this library is also optimized for small-size environments and only has the 2-size font loaded, while the 4-size (and larger) fonts are RLE to save space and rendered only if needed. 
I'm now looking for improved performance, space for fonts is not a problem.  My perf problem is that the WIO is too long busy updating TFT. 
A simple solution (for me) would be to have the 4-font loaded, or even just the numbers 0-9 preloaded.  Or cached or whatever.
Is there any way for a library user to do this? 
Below an extract of the section of the .h file that defines size 2 and 4 fonts.
Thanks for any advise

```
// Only load the fonts defined in User_Setup.h (to save space)
// Set flag so RLE rendering code is optionally compiled
#ifdef LOAD_GLCD
#include <Fonts/glcdfont.c>
#endif

#ifdef LOAD_FONT2
#include <Fonts/Font16.h>
#endif

#ifdef LOAD_FONT4
#include <Fonts/Font32rle.h>
#define LOAD_RLE
#endif
```

**Comment by Bodmer** [comment]
The library is not optimised for the WIO so uses the slower Arduino core SPI calls. Font rendering is already optimised and caching is not supported. Due to the way RLE is compressed caching will have a tiny effect. The performance is likely dominated by SPI transaction speed.

On an ESP32 a font 4 character is rendered with background in less than  0.3ms (SPI) or 0.09ms (8 bit parallel) which is faster that the TFT frame refresh period of circa15ms.

Run the TFT_Char_times sketch and report the times your are getting.
What SPI clock speed are you using?
What screen size are you updating?

**Reply by VeryOldFrank** [reply]
Hello, 

Thanks for your answer !
The WIO is standard equipped with a 320 x 240 screen and a SAMD51-based microcontroller that runs at 120MHz. 
I tested writing speed and found following: 
Writing "1234567890" in size 2 (without background) takes 2,07ms
Writing "1234567890" in size 4 (without background) takes 5.33ms
Drawing a filled rectangle that completely masks the "1234567890" in size 4 takes 5.15 ms

I'm looking for a way to speed up the size4 font writing so that the CPU would not be "fully taken" by this tft updating and could be doing other things (like processing incoming serial data that is going to be displayed on screen)

Code that I used for testing: 

```
#include "TFT_eSPI.h"      
TFT_eSPI tft;    //init library

unsigned long time1;
unsigned long time2;
unsigned long time3;
unsigned long time4;

void setup() {
  Serial.begin(115200);
  tft.begin();
  //tft.setRotation(r); //0-3  0: Button L-boven
  tft.setRotation(3); //0-3  USB kabel onderaan
  digitalWrite(LCD_BACKLIGHT, HIGH);  //LOW or HIGH
  tft.fillScreen(TFT_GREEN);
  time1 = millis();
  for (int i=1;i<101;i++) {
    tft.drawString("1234567890", 1,  1 , 2);
  }
  time2 = millis();
  for (int i=1;i<101;i++) {
    tft.drawString("1234567890", 1, 1 , 4);
  }
  time3 = millis();
  for (int i=1;i<101;i++) {
    tft.fillRect(1, 1, 145, 26, TFT_CYAN);
  }
  time4 = millis();
  Serial.print("Time1: ");Serial.println(time1);
  Serial.print("Time2: ");Serial.print(time2); Serial.print(" Delta: ");Serial.println(time2-time1);
  Serial.print("Time3: ");Serial.print(time3); Serial.print(" Delta: ");Serial.println(time3-time2); 
  Serial.print("Time4: ");Serial.print(time4); Serial.print(" Delta: ");Serial.println(time4-time3);  
}

void loop() {
  // put your main code here, to run repeatedly:
}
```

**Comment by Bodmer** [comment]
The maximum SPI clock rate for the SAMD51 is 24MHz and the ILI9341 will work at that rate, so set the frequency in your setup file to that value.

Online documents indicate there is a CPU clock boost mode that increases the CPU clock from 120MHz to 200 MHz but I do not know if this is supported by the Arduino environment or whether this impacts the maximum SPI clock rate.

**Comment by Bodmer** [comment]
The fillRect in the sketch completes in 5ms  this indicates a 12 Mbps data rate. It may be that you are already using a 24MHz SPI clock and that the Arduino board library SPI calls are slowing the achievable rate to 12MHz, in that case the only option is to try boosting the CPU clock.

**Comment by VeryOldFrank** [comment]
Hi, 
Thanks for all suggestions !
Here is what I found: 
a) Boosting the CPU clock to 200MHz: It seems that this is not something that can be done easily on the Arduino platform. I find no existing code to do this. I also think that I might run into other problems: timers, heat and battery life to name a few. So abandon this track. 
b) Running SPI at 24MHz: Interesting idea, The line of code that should get this done would seemingly be:
`SPI.beginTransaction(SPISettings(24000000, MSBFIRST, SPI_MODE0));`
Unfortunately, I find no way to get my program to compile correctly as soon as this line is in the code... I'll post a question on the Arduino forum for this. 
c) I ran some more display speed tests, for example: I replaced the fillRect by drawPixel inside 2 for loops painting one pixel at a time - that proved to be 10 times slower than fillRect. 
So overall conclusion is that we are working at max performance and getting SPI to work at 24MHz seems the only remaining track worth looking into.
